File: src/main/java/br/com/dusty/dkits/util/ItemStackUtils.java
package br.com.dusty.dkits.util;

import br.com.dusty.dkits.util.text.Text;
import org.bukkit.Color;
import org.bukkit.DyeColor;
import org.bukkit.Material;
import org.bukkit.enchantments.Enchantment;
import org.bukkit.inventory.ItemStack;
import org.bukkit.inventory.meta.ItemMeta;
import org.bukkit.inventory.meta.LeatherArmorMeta;
import org.bukkit.inventory.meta.PotionMeta;
import org.bukkit.material.Dye;
import org.bukkit.potion.PotionData;
import org.bukkit.potion.PotionType;

/**
 * Criação/personalização de {@link org.bukkit.inventory.ItemStack}
 */
public class ItemStackUtils {
	
	public static ItemStack rename(ItemStack itemStack, String name) {
		if(itemStack != null){
			ItemMeta itemMeta = itemStack.getItemMeta();
			if(itemMeta != null)
				itemMeta.setDisplayName(name);
			
			itemStack.setItemMeta(itemMeta);
		}
		
		return itemStack;
	}
	
	public static ItemStack enchant(ItemStack itemStack, int level, Enchantment... enchantments) {
		if(itemStack != null){
			for(Enchantment enchantment : enchantments)
				itemStack.addUnsafeEnchantment(enchantment, level);
		}
		
		return itemStack;
	}
	
	public static ItemStack color(ItemStack itemStack, Color color) {
		if(itemStack != null){
			if(itemStack.getType().name().startsWith("LEATHER_")){
				LeatherArmorMeta leatherArmorMeta = (LeatherArmorMeta) itemStack.getItemMeta();
				leatherArmorMeta.setColor(color);
				
				itemStack.setItemMeta(leatherArmorMeta);
			}
		}
		
		return itemStack;
	}
	
	public static ItemStack dye(DyeColor c) {
		Dye dye = new Dye();
		dye.setColor(c);
		
		return dye.toItemStack(1);
	}
	
	public static ItemStack potions(int amount, boolean extended, boolean upgraded, PotionType potionType, boolean splash) {
		ItemStack itemStack = new ItemStack(Material.POTION, amount);
		
		PotionData potionData = new PotionData(potionType, extended, upgraded);
		
		PotionMeta potionMeta = (PotionMeta) itemStack.getItemMeta();
		potionMeta.setBasePotionData(potionData);
		
		itemStack.setItemMeta(potionMeta);
		
		return itemStack;
	}
	
	public static ItemStack setDescription(ItemStack itemStack, String description) {
		ItemMeta itemMeta = itemStack.getItemMeta();
		itemMeta.setLore(StringUtils.fancySplit(description, 32));
		
		itemStack.setItemMeta(itemMeta);
		
		return itemStack;
	}
	
	/**
	 * Retorna o <b>'displayName'</b> não-formatado de uma {@link ItemStack}, se houver, ou o <b>name()</b> de seu {@link org.bukkit.Material}, caso contrário.
	 *
	 * @param itemStack
	 * @return <b>'displayName'</b> não-formatado de uma {@link ItemStack}, se houver, ou o <b>name()</b> de seu {@link org.bukkit.Material}, caso contrário.
	 * Pode, ainda, retornar <b>null</b> se a {@link ItemStack} for 'null'.
	 */
	public static String getUnformattedDisplayName(ItemStack itemStack) {
		String displayName = null;
		
		if(itemStack != null){
			ItemMeta itemMeta = itemStack.getItemMeta();
			if(itemMeta != null && itemMeta.hasDisplayName())
				displayName = itemMeta.getDisplayName();
			else
				displayName = itemStack.getType().name();
		}
		
		return Text.clearFormatting(displayName);
	}
}

File: src/main/kotlin/br/com/dusty/dkits/util/ItemStackUtils.kt
package br.com.dusty.dkits.util

import br.com.dusty.dkits.util.text.Text
import org.bukkit.Color
import org.bukkit.DyeColor
import org.bukkit.Material
import org.bukkit.enchantments.Enchantment
import org.bukkit.inventory.ItemStack
import org.bukkit.inventory.meta.LeatherArmorMeta
import org.bukkit.inventory.meta.PotionMeta
import org.bukkit.material.Dye
import org.bukkit.potion.PotionData
import org.bukkit.potion.PotionType

fun ItemStack.hasName(name: String): Boolean {
	val itemMeta = this.itemMeta
	if (itemMeta != null && itemMeta.hasDisplayName())
		return Text.clearFormatting(itemMeta.displayName) == name

	return false
}

fun ItemStack.rename(name: String): ItemStack? {
	val itemMeta = this.itemMeta
	if (itemMeta != null)
		itemMeta.displayName = name

	this.itemMeta = itemMeta

	return this
}

fun ItemStack.enchant(level: Int, vararg enchantments: Enchantment): ItemStack? {
	for (enchantment in enchantments)
		this.addUnsafeEnchantment(enchantment, level)

	return this
}

fun ItemStack.color(color: Color): ItemStack? {
	if (this.type.name.startsWith("LEATHER_")) {
		val leatherArmorMeta = this.itemMeta as LeatherArmorMeta
		leatherArmorMeta.color = color

		this.itemMeta = leatherArmorMeta
	}

	return this
}

fun ItemStack.setDescription(description: String): ItemStack {
	val itemMeta = this.itemMeta
	itemMeta.lore = description.fancySplit(32)

	this.itemMeta = itemMeta

	return this
}

/**
 * Retorna o **'displayName'** não-formatado de uma [ItemStack], se houver, ou o **name()** de seu [org.bukkit.Material], caso contrário.
 *
 * @param itemStack
 * @return **'displayName'** não-formatado de uma [ItemStack], se houver, ou o **name()** de seu [org.bukkit.Material], caso contrário.
 * Pode, ainda, retornar **null** se a [ItemStack] for 'null'.
 */
fun ItemStack.getUnformattedDisplayName(): String? {
	val itemMeta = this.itemMeta

	val displayName = if (itemMeta != null && itemMeta.hasDisplayName())
		itemMeta.displayName
	else
		this.type.name

	return Text.clearFormatting(displayName)
}

/**
 * Criação/personalização de [org.bukkit.inventory.ItemStack]
 */
object ItemStackUtils {

	fun dye(c: DyeColor): ItemStack {
		val dye = Dye()
		dye.color = c

		return dye.toItemStack(1)
	}

	fun potions(amount: Int, extended: Boolean, upgraded: Boolean, potionType: PotionType, splash: Boolean): ItemStack {
		val itemStack = ItemStack(Material.POTION, amount)

		val potionData = PotionData(potionType, extended, upgraded)

		val potionMeta = itemStack.itemMeta as PotionMeta
		potionMeta.basePotionData = potionData

		itemStack.itemMeta = potionMeta

		return itemStack
	}
}
